### PR TITLE
Massive cleanup better auth

### PR DESCRIPTION
**Pull request by reifcode**
Cleaned up code, with a more manageable auth flow, and fixed tests.

### CODE REVIEW

File: auth.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+	Authenticate(signer *Signer, request *http.Request, context ScopeContext, feed Feed) error
+}
+
+type applicationAuthenticator struct {
+	key    string
+	secret string
+}
+
+func (a applicationAuthenticator) Authenticate(signer *Signer, request *http.Request, context ScopeContext, feed Feed) error {
+	return signer.SignHTTP(request)
```
Comment (JelteF) [inline auth.go]: We want to deprecate http signatures, so this is shouldn't be needed. It's better to use a JWT token with the `"*"` feed permissions.

File: feed.go [hunk 2]
```diff
@@ -253,7 +253,7 @@ func (f *baseFeed) GetFollowings(limit int, offset int) ([]FeedID, error) {
 		outputFeeds []FeedID
 	)
 
-	endpoint := "feed/" + f.FeedSlug + "/" + f.UserID + "/" + "following" + "/"
+	endpoint := "feed/" + f.FeedSlug + "/" + f.UserID + "/" + "follows" + "/"
```
Comment (JelteF) [inline feed.go]: This change seems like a mistake. Maybe the one below as well?

File: feed_flat_test.go [hunk 3]
```diff
@@ -8,16 +8,18 @@ import (
 	getstream "github.com/GetStream/stream-go"
 )
 
-func getFlatFeed(client *getstream.Client) *getstream.FlatFeed {
-	f, _ := client.FlatFeed("flat", RandString(8))
-	return f
+func getFlatFeed(client *getstream.Client) (*getstream.FlatFeed, error) {
+	return client.FlatFeed("flat", RandString(8))
```
Comment (JelteF) [inline feed_flat_test.go]: I would pass `t` and do `t.Fatal` here if there's an error. Causes a lot less error checking clutter in the tests.

File: signer.go [hunk 4]
```diff
@@ -39,60 +42,37 @@ func (s Signer) GenerateToken(message string) string {
...
-		// "iss":
-		// "nbf": time.Now().Unix(),
-		// "sub":
-	}
-
-	if feedIDWithoutColon != "" {
-		claims["feed_id"] = feedIDWithoutColon
-	} else {
-		claims["feed_id"] = "*"
+func (s Signer) SignHTTP(request *http.Request) error {
```
Comment (JelteF) [inline signer.go]: Like I said above. Better just to throw this out and only use JWT for authentication.

File: feed.go (hunk 2 above)
Comment (reifcode) [inline feed.go]: According to the REST docs it's correct, isn't it? https://getstream.io/docs_rest/#follows

File: signer.go (hunk 4 above)
Comment (reifcode) [inline signer.go]: not strictly related, but shouldn't this be reflected in the REST documentation?

File: auth.go (hunk 1 above)
Comment (JelteF) [inline auth.go]: Ok I was wrong about this.

File: context.go [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+		context: ScopeContextNoContext,
+		auth:    ApplicationAuthentication,
+	},
+	{
+		re:      regexp.MustCompile("^activities/$"),
+		context: ScopeContextActivities,
+		auth:    FeedAuthentication,
+	},
+	{
+		re:      regexp.MustCompile("^feed/.+/.+/followers/$"),
```
Comment (JelteF) [inline context.go]: don't use `.+` in the matches, use `[^/]+`, otherwise paths are eaten as well.

File: feed.go (hunk 2 above)
Comment (reifcode) [inline feed.go]: The docs document only `follows/` so it's better to keep the implementation consistent with the documentation.

File: signer.go (hunk 4 above)
Comment (reifcode) [inline signer.go]: See above

**Pull request merged**